viamonvifdiscovery/service.go

```diff
@@ -0,0 +1,201 @@
+	Username   string `json:"username"`
+	Password   string `json:"password"`
+	credNumber int
+}
+
+// Config is the config for the discovery service.
+type Config struct {
+	Credentials []Creds `json:"credentials"`
+}
+
+// Validate validates the discovery service.
+func (cfg *Config) Validate(_ string) ([]string, error) {
+	// check that all creds have both usernames and passwords set. Note a credential can have both fields empty
+	for _, cred := range cfg.Credentials {
+		if cred.Username != "" && cred.Password == "" {
+			return nil, fmt.Errorf("credential %v missing password", cred.Username)
+		}
+		if cred.Password != "" && cred.Username == "" {
+			return nil, fmt.Errorf("credential missing username, has password %v", cred.Password)
+		}
+	}
+	return nil, nil
+}
+
+type rtspDiscovery struct {
+	resource.Named
+	resource.AlwaysRebuild
+	Credentials []Creds
+	logger      logging.Logger
+}
+
+func newDiscovery(_ context.Context, _ resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (discovery.Service, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+	dis := &rtspDiscovery{
+		Named: conf.ResourceName().AsNamed(),
+		// place the empty credentials first, so we find the insecure cameras before other cameras.
+		Credentials: append([]Creds{emptyCred}, cfg.Credentials...),
+		logger:      logger,
+	}
+
+	dis.setCredNumbers()
+	return dis, nil
+}
+
+// since camera logins can have the same username,
+// check how many repeated usernames we have and index them, to ensure cameras have unique names.
+func (dis *rtspDiscovery) setCredNumbers() {
+	// usernames do not have to be unique, so we want to add an additional label to ensure cameras have unique names.
+	for index, cred := range dis.Credentials {
+		for counter, otherCreds := range dis.Credentials {
+			if counter <= index {
+				continue
+			}
+			// increase the credNumber for later cameras.
+			if cred.Username == otherCreds.Username {
+				dis.Credentials[counter].credNumber++
+			}
+		}
+	}
+}
+
+// Close closes the discovery service.
+func (dis *rtspDiscovery) Close(_ context.Context) error {
+	return nil
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *rtspDiscovery) DiscoverResources(_ context.Context, _ map[string]any) ([]resource.Config, error) {
+	potentialCams := []resource.Config{}
+	insecureURLs := []string{}
+
+	for _, cred := range dis.Credentials {
+		// if we have multiple empty/insecure credentials configured for some reason, skip to the next credential.
+		// only skip if we have not processed insecure credentials yet.
+		if cred.isInsecure() && len(insecureURLs) > 0 {
+			continue
+		}
+
+		list, err := viamonvif.DiscoverCameras(cred.Username, cred.Password, dis.logger, nil)
+		if err != nil {
+			return nil, err
+		}
+		cameraNumber := 0
+		for _, l := range list.Cameras {
+			dis.logger.Debugf("%s %s %s", l.Manufacturer, l.Model, l.SerialNumber)
```

Comment:
I'd rather use RTSPAddress nomenclature rather than "cameras" since that's the attribute we discover to populate a viamrtsp camera